File: typescript-sdk/src/connectors/openai.ts
/**
 * OpenAI connector with automatic RAG tracing - TypeScript SDK
 */

import {
  OpenAIMessage,
  OpenAICompletionRequest,
  OpenAICompletionResponse,
  ChunkData
} from '../types';
import { tracer } from '../tracer';

/**
 * Traced OpenAI Chat Completion function
 * Drop-in replacement for openai.chat.completions.create()
 */
export async function chatCompletion(
  params: OpenAICompletionRequest,
  openaiClient?: any
): Promise<OpenAICompletionResponse> {
  const userMessage = params.messages.find((m: OpenAIMessage) => m.role === 'user');
  const userInput = userMessage?.content || 'OpenAI Chat Completion';

  return tracer.startTrace(userInput, async (ctx) => {
    const startTime = Date.now();

    try {
      // Add prompt to trace
      const fullPrompt = params.messages.map((m: OpenAIMessage) => `${m.role}: ${m.content}`).join('\n');
      ctx.addPrompt(fullPrompt);

      let response: OpenAICompletionResponse;

      if (openaiClient) {
        // Use provided client
        response = await openaiClient.chat.completions.create(params);
      } else {
        // Try to use global OpenAI client if available
        try {
          // @ts-ignore - OpenAI might be available globally
          const OpenAI = require('openai');
          const client = new OpenAI();
          response = await client.chat.completions.create(params);
        } catch (error) {
          throw new Error('OpenAI client not provided and global OpenAI not available. Install openai package or pass client.');
        }
      }

      // Extract response data
      const output = response.choices[0]?.message?.content || '';
      const tokensIn = response.usage?.prompt_tokens;
      const tokensOut = response.usage?.completion_tokens;
      const latencyMs = Date.now() - startTime;

      ctx.setModelOutput(output, params.model, tokensIn, tokensOut);

      return response;
    } catch (error) {
      ctx.setError(error instanceof Error ? error.message : String(error));
      throw error;
    }
  }) as Promise<OpenAICompletionResponse>;
}

/**
 * Wrap an existing OpenAI client with automatic tracing
 */
export function wrapOpenAIClient(client: any): any {
  const originalCreate = client.chat.completions.create.bind(client.chat.completions);

  client.chat.completions.create = async (params: OpenAICompletionRequest) => {
    return chatCompletion(params, { chat: { completions: { create: originalCreate } } });
  };

  return client;
}

/**
 * Create a traced OpenAI client factory
 */
export function createTracedOpenAI(OpenAI: any, options?: any): any {
  const client = new OpenAI(options);
  return wrapOpenAIClient(client);
}

/**
 * RAG-specific utilities for OpenAI
 */
export const openaiUtils = {
  /**
   * Helper to add retrieved chunks before OpenAI call
   */
  async chatCompletionWithRAG(
    params: OpenAICompletionRequest,
    retrievedChunks: ChunkData[],
    scores?: number[],
    openaiClient?: any
  ): Promise<OpenAICompletionResponse> {
    const userMessage = params.messages.find((m: OpenAIMessage) => m.role === 'user');
    const userInput = userMessage?.content || 'OpenAI RAG Query';

    return tracer.startTrace(userInput, async (ctx) => {
      // Add retrieved chunks to trace
      ctx.addRetrievedChunks(retrievedChunks, scores);

      // Modify the prompt to include context
      const context = retrievedChunks.map(chunk => chunk.text).join('\n\n');
      const modifiedMessages = [...params.messages];
      
      // Find the last user message and enhance it with context
      const lastUserMsgIndex = modifiedMessages.map(m => m.role).lastIndexOf('user');
      if (lastUserMsgIndex !== -1) {
        modifiedMessages[lastUserMsgIndex] = {
          ...modifiedMessages[lastUserMsgIndex],
          content: `Context:\n${context}\n\nQuestion: ${modifiedMessages[lastUserMsgIndex].content}`
        };
      }

      const modifiedParams = { ...params, messages: modifiedMessages };
      
      // Call the regular traced completion
      return chatCompletion(modifiedParams, openaiClient);
    }) as Promise<OpenAICompletionResponse>;
  }
};

// Export types for convenience
export type { OpenAIMessage, OpenAICompletionRequest, OpenAICompletionResponse }; 

File: ragtoolkit/sdk/connectors/openai.py
"""
OpenAI Connector for RAG Toolkit

Wraps OpenAI client with automatic tracing capabilities.
"""

import functools
import time
from typing import Any, Dict, Optional, Union, List
from ..tracer import RAGTracker


def with_tracing(func):
    """Decorator to add tracing to OpenAI API calls"""
    @functools.wraps(func)
    def wrapper(self, *args, **kwargs):
        # Extract common parameters
        messages = kwargs.get('messages', args[0] if args else [])
        model = kwargs.get('model', getattr(self, 'model', 'gpt-3.5-turbo'))
        
        # Start tracing
        tracker = RAGTracker()
        
        # Extract user input from messages
        user_input = None
        if messages and isinstance(messages, list):
            user_messages = [m for m in messages if m.get('role') == 'user']
            if user_messages:
                user_input = user_messages[-1].get('content', '')
        
        with tracker.trace_context(user_input=user_input or 'OpenAI API call') as trace:
            # Add system/user prompts
            if messages:
                full_prompt = "\n".join([f"{m.get('role', 'unknown')}: {m.get('content', '')}" 
                                       for m in messages])
                tracker.add_prompt(full_prompt)
            
            # Record start time
            start_time = time.time()
            
            try:
                # Make the original API call
                response = func(self, *args, **kwargs)
                
                # Record timing
                end_time = time.time()
                latency_ms = int((end_time - start_time) * 1000)
                
                # Extract response content and tokens
                if hasattr(response, 'choices') and response.choices:
                    content = response.choices[0].message.content
                    tracker.set_model_output(
                        output=content,
                        model_name=model,
                        tokens_in=getattr(response.usage, 'prompt_tokens', None) if hasattr(response, 'usage') else None,
                        tokens_out=getattr(response.usage, 'completion_tokens', None) if hasattr(response, 'usage') else None,
                        latency_ms=latency_ms
                    )
                
                return response
                
            except Exception as e:
                # Record error
                tracker.set_model_output(
                    output=f"Error: {str(e)}",
                    model_name=model,
                    latency_ms=int((time.time() - start_time) * 1000)
                )
                raise
    
    return wrapper


class ChatCompletion:
    """
    Drop-in replacement for OpenAI ChatCompletion with automatic tracing.
    
    Usage:
        from ragtoolkit.connectors.openai import ChatCompletion
        from openai import OpenAI
        
        client = OpenAI()
        # Replace: response = client.chat.completions.create(...)
        response = ChatCompletion.create(client, messages=[...], model="gpt-4")
    """
    
    @staticmethod
    @with_tracing
    def create(client, **kwargs):
        """Create a chat completion with automatic tracing"""
        return client.chat.completions.create(**kwargs)
    
    @classmethod 
    def wrap_client(cls, client):
        """
        Wrap an existing OpenAI client to add tracing to all chat completions.
        
        Args:
            client: OpenAI client instance
            
        Returns:
            Modified client with tracing enabled
        """
        original_create = client.chat.completions.create
        client.chat.completions.create = with_tracing(original_create)
        return client


def traced_openai_client(api_key: Optional[str] = None, **kwargs):
    """
    Create a new OpenAI client with tracing enabled.
    
    Args:
        api_key: OpenAI API key
        **kwargs: Additional arguments for OpenAI client
        
    Returns:
        OpenAI client with automatic tracing
    """
    try:
        from openai import OpenAI
    except ImportError:
        raise ImportError("openai package is required. Install with: pip install openai")
    
    client = OpenAI(api_key=api_key, **kwargs)
    return ChatCompletion.wrap_client(client) 